src/hydranfc/hydranfc_bbio_reader.c

```diff
@@ -0,0 +1,199 @@
+ * limitations under the License.
+ */
+
+#include "common.h"
+#include "tokenline.h"
+#include "trf797x.h"
+#include "hydrabus_bbio.h"
+#include "hydranfc_bbio_reader.h"
+
+#include <string.h>
+
+#define CMD_SUCCESS                 0x00
+#define SET_MODE_ISO_14443A_FAILURE 0x01 // "Error ISO Control Register read=0x%02lX (should be 0x88)\r\n"
+
+#define RF_OFF                    0x0001
+#define RF_ON                     0x0002
+#define CMD_SEND_BITS             0x0003
+#define CMD_SEND_BYTES            0x0004
+#define SET_MODE_ISO_14443A       0x0005
+#define SET_MODE_ISO_15693        0x0006
+#define CMD_EXIT                  0xDEAD
+
+bool init_gpio(t_hydra_console *con);
+
+bool deinit_gpio(void);
+
+uint8_t set_mode_iso_14443A(void) {
+    uint8_t data_buf[2];
+
+    /* Test ISO14443-A/Mifare read UID */
+    Trf797xInitialSettings();
+    Trf797xResetFIFO();
+
+    /*
+     * Write Modulator and SYS_CLK Control Register (0x09) (13.56Mhz SYS_CLK
+     * and default Clock 13.56Mhz))
+     */
+    data_buf[0] = MODULATOR_CONTROL;
+    data_buf[1] = 0x31;
+    Trf797xWriteSingle(data_buf, 2);
+
+    /*
+     * Configure Mode ISO Control Register (0x01) to 0x88 (ISO14443A RX bit
+     * rate, 106 kbps) and no RX CRC (CRC is not present in the response))
+     */
+    data_buf[0] = ISO_CONTROL;
+    data_buf[1] = 0x88;
+    Trf797xWriteSingle(data_buf, 2);
+
+    data_buf[0] = ISO_CONTROL;
+    Trf797xReadSingle(data_buf, 1);
+
+    if (data_buf[0] != 0x88) {
+        return SET_MODE_ISO_14443A_FAILURE;
+    }
+
+    return CMD_SUCCESS;
+}
+
+void set_mode_iso_15693(void) {
+    static uint8_t data_buf[2];
+
+    /* Test ISO15693 read UID */
+    Trf797xInitialSettings();
+    Trf797xResetFIFO();
+
+    /* Write Modulator and SYS_CLK Control Register (0x09) (13.56Mhz SYS_CLK and default Clock 13.56Mhz)) */
+    data_buf[0] = MODULATOR_CONTROL;
+    data_buf[1] = 0x31;
+    Trf797xWriteSingle(data_buf, 2);
+
+    /* Configure Mode ISO Control Register (0x01) to 0x02 (ISO15693 high bit rate, one subcarrier, 1 out of 4) */
+    data_buf[0] = ISO_CONTROL;
+    data_buf[1] = 0x02;
+    Trf797xWriteSingle(data_buf, 2);
+}
+
+void get_cmd(t_hydra_console *con, uint16_t *tag, uint32_t *length, uint8_t *data) {
+    uint8_t c;
+    *tag = 0;
+    chnRead(con->sdu, &c, 1);
+
+    if (c == 0) {
+        cprint(con, "BBIO2", 5);
+        // Get 0xC Command
+        chnRead(con->sdu, &c, 1); // 0xC
+
+        cprint(con, "NFC", 3);
+
+        // Get Tag First Byte
+        chnRead(con->sdu, &c, 1);
+    }
+    *tag = c;
+
+    chnRead(con->sdu, &c, 1);
+    *tag += (c << 8);
+
+    chnRead(con->sdu, (uint8_t *) length, 4);
+
+    if (*length > 0) {
+        chnRead(con->sdu, data, *length);
+    }
+}
+
+void send_response(t_hydra_console *con, uint16_t tag, uint32_t length, uint8_t *data) {
+
+    cprint(con, (char *) &tag, 2);
+    cprint(con, (char *) &length, 4);
+    if (length > 0) {
+        cprint(con, (char *) data, length);
+    }
```

Comment:
You should not send user-supplied data back, but only command status and/or data read from the protocol.